### PR TITLE
Pull Request for #63: Make current dependencies optional

### PR DESCRIPTION
**Pull request by yuchuan2016**
@stanfordquan , can you take a look at this? I'm a little stuck on how to specify the required packages for each build. Previously if we add a new package requirement, we just add it in the argument of `check_xx` function in `configuration_test.py`. But now the check is after the definition of the builders. Also I'm not sure how to pass the packages arguments in a `dict` syntax.

My current solution is to store the required packages in `config_global.yaml` and add some "hard code" in `gslab_python`. Can you think of a better solution of this? I think this problem won't exist in a world with automatic dependencies installation.

### CODE REVIEW

**Comment by yuchuan2016** [comment]
@stanfordquan , please should hold on for this pull request per discussion.

**Comment by yuchuan2016** [comment]
@stanfordquan , can you take a look at this (or @arosenbe if you have time)? I have created three scripts in a top-level directory `config`, which should be run manually and separately. I move the `configuration_test` in `config`, and rename it as `configuration`, since it does not have many tests now. It basically gets mode, release version and cache_dir, and checks git-lfs and lyx.

In `SConstruct`, I first call `configuration` to get `mode, vers, cache_dir`, then define the builder, and then load stata executable if the stata builder is in the environment. No checks for Stata/R/Python and no checks for the packages now.

Tell me if you have any suggestions, or you think there is something that we need to ask Matt.

**Comment by arosenbe** [comment]
I'll look @stanfordquan

File: config/config_python.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+        pip.main(['install', 'git+http://git@github.com/gslab-econ/gslab_python.git@4.1.0'])
+
+    # Install required packages using pip
+    packages = {'pyyaml': True}
+    installed_packages = [pkg.key for pkg in pip.get_installed_distributions()]
+    for pkg in packages.keys():
+        if pkg not in installed_packages or packages[pkg]:
+            pip.main(['install', '--upgrade', pkg])
+
+def check_gslab_tools(gslab_python_version):
```
Comment (arosenbe) [inline config/config_python.py]: This makes a pretty strong promise about backwards compatibility. Maybe we just force installation of the current version?

File: config/config_python.py [hunk 2]
```diff
@@ -0,0 +1,34 @@
+import pip
+import pkg_resources
+
+def main():
+    # Install gslab_tools if it's not installed yet or it has the wrong version
+    if not check_gslab_tools('4.1.0'):
+        pip.main(['install', 'git+http://git@github.com/gslab-econ/gslab_python.git@4.1.0'])
```
Comment (arosenbe) [inline config/config_python.py]: We'll have to update this line for after the release. I think we can drop a tag in for the branch name [here](https://coderwall.com/p/-wbo5q/pip-install-a-specific-github-repo-tag-or-branch), which would help keep things reproducible.

File: config/config_r.r [hunk 3]
```diff
@@ -0,0 +1,28 @@
+Main <- function(){
```
Comment (arosenbe) [inline config/config_r.r]: Any reason this is capitalized?

File: config/config_r.r [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+    CRAN_packages   <- c("yaml", "ggplot2")
+    GitHub_packages <- NULL
+    # GitHub packages should be the full path to the relevant repo
+    if (!is.null(GitHub_packages)) {
```
Comment (arosenbe) [inline config/config_r.r]: Maybe a comment explaining this block?

File: config/config_r.r [hunk 5]
```diff
@@ -0,0 +1,28 @@
+Main <- function(){
+    CRAN_packages   <- c("yaml", "ggplot2")
+    GitHub_packages <- NULL
+    # GitHub packages should be the full path to the relevant repo
```
Comment (arosenbe) [inline config/config_r.r]: Move comment one line above or to same line as `GitHub_packages <- NULL`.

File: config/config_python.py (hunk 1 above)
Comment (yuchuan2016) [reply config/config_python.py]: Do you mean `gslab_tools` or the other packages?

File: config/config_python.py (hunk 2 above)
Comment (yuchuan2016) [reply config/config_python.py]: I agree.

**Comment by yuchuan2016** [comment]
@arosenbe , I have updated the template per your comments. I keep the switch between lyx and latex and the corresponding checks. Can you take another look at my changes? Thanks!

File: config/config_python.py (hunk 1 above)
Comment (arosenbe) [reply config/config_python.py]: I guess both, but I'm mostly concerned about `gslab_tools`.

Comment (yuchuan2016) [reply config/config_python.py]: So we do not check the revision any more, but force the users to install what's currently in `master`?

File: config/config_r.r [hunk 6]
```diff
@@ -1,7 +1,10 @@
-Main <- function(){
-    CRAN_packages   <- c("yaml", "ggplot2")
-    GitHub_packages <- NULL
-    # GitHub packages should be the full path to the relevant repo
+# Required packages installed from CRAN.
+CRAN_packages   <- c("yaml")
+# Required packages installed from Github. Should be the full path to the relevant repo.
+GitHub_packages <- NULL
+
+main <- function(CRAN_packages = NULL, GitHub_packages = NULL) {
```
Comment (arosenbe) [inline config/config_r.r]: We might want to add arguments to control the install_CRAN options for repo, dependency, and quiet. We'll set their defaults to the specs you've already filled in.

File: config/config_python.py (hunk 1 above)
Comment (arosenbe) [reply config/config_python.py]: Yes. Except we should force installation of the v4.1.0 release.

```python
pip.main(['install', '--upgrade', 
          'git+http://git@github.com/gslab-econ/gslab_python.git@v4.1.0'])
```

File: config/config_r.r (hunk 6 above)
Comment (yuchuan2016) [reply config/config_r.r]: Another question is, should we force users to re-install the packages if they already install the package before? We might want to make this consistent between python, r and stata.

Comment (arosenbe) [reply config/config_r.r]: I'd vote no. We don't force a re-installation unless we need to specify a specific version as with gslab_python. I recognize that this creates an inconsistency between python/R and stata, but I'm fine with that, unless you know an easy way to check if a stata package has been installed.

**Comment by yuchuan2016** [comment]
@arosenbe , one concern is that, suppose users are collaborating on one project, and they have all run the `config` scripts. Then one user add some dependencies in the `config` scripts without notifying others, which is very likely. Then other users will not run the `config` scripts again, because they think they have already run these scripts before. Then the build will fail.

Do you think this is a problem? Should we make running the config scripts part of the scons build?

**Comment by arosenbe** [comment]
@yuchuan2016, 

MG didn't like the idea of including the config scripts in the scons build, one more thing to break. I'm willing to just take him at his word. 

The current practice is to enumerate dependencies on external packages in the readme, which suffers the same problem. I think it suffices to mention these scripts in the FAQ. We might have another section there called "why did my build fail," where we can mention common errors like this one.

**Comment by yuchuan2016** [comment]
@arosenbe , great! Shall you take a look to see there is any other problem? Thanks!

File: config/config_python.py [hunk 7]
```diff
@@ -1,37 +1,20 @@
...
-packages = {'pyyaml': True}
+# The collection of required packages
+packages = ['pyyaml']
 
-def main(packages):
+def main(packages, upgrade = False):
     # Install gslab_tools if it's not installed yet or it has the wrong version
-    if not check_gslab_tools('4.1.0'):
-        pip.main(['install', 'git+http://git@github.com/gslab-econ/gslab_python.git@4.1.0'])
+    pip.main(['install', 'git+http://git@github.com/gslab-econ/gslab_python.git@4.1.0'])
```
Comment (arosenbe) [inline config/config_python.py]: I think we want the `--upgrade` flag here too. Also, we normally release with the tag `v4.1.0`; so change that too.

File: config/config_r.r [hunk 8]
```diff
@@ -3,7 +3,8 @@ CRAN_packages   <- c("yaml")
 # Required packages installed from Github. Should be the full path to the relevant repo.
 GitHub_packages <- NULL
 
-main <- function(CRAN_packages = NULL, GitHub_packages = NULL) {
+main <- function(CRAN_packages = NULL, GitHub_packages = NULL,
+                 CRAN_dependency = TRUE, CRAN_quiet = TRUE, upgrade = FALSE) {
```
Comment (arosenbe) [inline config/config_r.r]: I think we should allow a pre-filled `repo` argument as well.

**Comment by yuchuan2016** [comment]
@arosenbe , I implemented your comments. Going to merge to 4.1.0

**Pull request merged**